include/cuco/detail/probe_sequences.cuh

```diff
@@ -0,0 +1,310 @@
+ *
+ * @tparam Key Type used for keys
+ * @tparam Value Type of the mapped values
+ * @tparam CGSize Number of threads in CUDA Cooperative Groups
+ * @tparam Hash1 Unary callable type
+ * @tparam Hash2 Unary callable type
+ * @tparam Scope The scope in which multimap operations will be performed by
+ * individual threads
+ */
+template <typename Key,
+          typename Value,
+          uint32_t CGSize          = 8,
+          typename Hash1           = cuco::detail::MurmurHash3_32<Key>,
+          typename Hash2           = cuco::detail::MurmurHash3_32<Key>,
+          cuda::thread_scope Scope = cuda::thread_scope_device>
+class double_hashing : public probe_sequence_base<Key, Value, CGSize, Scope> {
+ public:
+  using probe_sequence_base_type = probe_sequence_base<Key, Value, CGSize, Scope>;
+  using iterator                 = typename probe_sequence_base_type::iterator;
+  using probe_sequence_base_type::capacity_;
+  using probe_sequence_base_type::cg_size;
+  using probe_sequence_base_type::slots_;
+  using probe_sequence_base_type::uses_vector_load;
+  using probe_sequence_base_type::vector_width;
+
+  /**
+   * @brief Constructs a double hashing scheme based on the given hash map features.
+   *
+   * @param slots Pointer to beginning of the hash map slots
+   * @param capacity Capacity of the hash map
+   * @param hash1 First hasher to hash each key
+   * @param hash2 Second hasher to determine step size
+   */
+  __host__ __device__ explicit double_hashing(iterator slots,
+                                              std::size_t capacity,
+                                              Hash1 hash1 = Hash1{},
+                                              Hash2 hash2 = Hash2{}) noexcept
+    : probe_sequence_base<Key, Value, CGSize, Scope>{slots, capacity}, hash1_{hash1}, hash2_{hash2}
+  {
+  }
+
+  /**
+   * @brief Returns the initial slot for a given key `k`.
+   *
+   * If vector-load is enabled, the return slot is always a multiple of (`cg_size` * `vector_width`)
+   * to avoid illegal memory access.
+   *
+   * @tparam CG CUDA Cooperative Groups type
+   * @param g the Cooperative Group for which the initial slot is needed
+   * @param k The key to get the slot for
+   * @return Pointer to the initial slot for `k`
+   */
+  template <typename CG>
+  __device__ iterator initial_slot(CG const& g, Key const k) noexcept
+  {
+    std::size_t index;
+    auto const hash_value = hash1_(k);
+    if constexpr (uses_vector_load()) {
+      // step size in range [1, prime - 1] * cg_size * vector_width
+      step_size_ = (hash2_(k) % (capacity_ / (cg_size() * vector_width()) - 1) + 1) * cg_size() *
+                   vector_width();
+      index = hash_value % (capacity_ / (cg_size() * vector_width())) * cg_size() * vector_width() +
+              g.thread_rank() * vector_width();
+    } else {
+      // step size in range [1, prime - 1] * cg_size
+      step_size_ = (hash2_(k) % (capacity_ / cg_size() - 1) + 1) * cg_size();
```

Comment:
@jrhemstad The current implementation is to accommodate non-arithmetic (or custom) key types. This is not ideal cause the secondary clustering issue still exists. 
```cpp
step_size_ = (hash2_(k + 1) % (capacity_ / cg_size() - 1) + 1) * cg_size();
```
Passing `k + 1` to the hasher instead can get rid of the secondary clustering issue but that requires the target key to be an arithmetic type or has `+` operator defined. So what do we would like to support? An algorithm working for all key types or one can avoid secondary clustering?

Reply:
I'm not following.

Why is `k +1` necessary for double hashing to eliminate clustering?

@sleeepyjack ?

Reply:
Right. Since we're using the same hasher for `h1` and `h2` we need this `+1` so eliminate secondary clustering effects.

Reply:
What if the hasher takes an optional seed and then we use two different seeds for double hashing?

Reply:
Instead of `hash2_(k +1)`, can't we do `hash2_(k) + 1`?

Reply:
If `hash(m)` and `hash(n)` are the same, `hash(m) + 1` and `hash(n) + 1` will be indentical as well. Thus this can not differentiate step sizes for `m` and `n`.

Reply:
So we're talking about hash collisions here? i.e., `m != n` but `hash(m) == hash(n)`?

How much are true hash collisions a problem? Typically I think the problem results more from collisions that result from `hash(m) != hash(n)` but `hash(m) % capacity == hash(n) % capacity` (I don't know what to call this kind of collision).

Reply:
Like you said, the easiest way is to mutate hash functions. Instead of using the same seed for both hashers in double hashing, we can make the first hasher take `seed = 0` and the second one takes `seed = 1`.

Reply:
> Typically I think the problem results more from collisions that result from hash(m) != hash(n) but hash(m) % capacity == hash(n) % capacity

Reply:
> Typically I think the problem results more from collisions that result from hash(m) != hash(n) but hash(m) % capacity == hash(n) % capacity

Reply:
> Typically I think the problem results more from collisions that result from `hash(m) != hash(n)` but `hash(m) % capacity == hash(n) % capacity`

Since our capacity is prime, the probing sequences should diverge after the first probing step.

The problem only occurs if both the initial slot and the step size are identical for two or more distinct keys.
The thing with secondary clustering is, that we might end up with regions in the map that are more dense than others and these regions tend to grow which increases the average probing length. 
The big advantage of double hashing is that iff the two hash functions are not the same and the capacity is prime, then we can virtually eliminate this effect entirely.

Reply:
> The problem only occurs if both the initial slot and the step size are identical for two or more distinct keys.

We can't totally prevent this though, right? Even with the old implementation, there is nothing to say that `hash(k) != hash(k+1)`. 

Can we reasonably claim that `hash(k) != hash(k+1)` will be true less often than `hash(k) != hash(k) + 1`?

Reply:
If we're talking about up-to 64bit keys then Murmur is a true permutation over the index domain 
and thus `hash(k) != hash(k+1)` holds. For larger keys there might be collisions though.

Reply:
Also, in multi-value scenarios we might get secondary clustering anyway due to key collisions.